pyanaconda/kickstart.py

```diff
@@ -1915,6 +1914,70 @@ def execute(self, *args):
             desktop.runlevel = 3
             desktop.write()
 
+class Snapshot(commands.snapshot.RHEL7_Snapshot):
+    def setup(self, storage, ksdata, instClass):
+        """ Prepare the snapshot.
+
+            This will also do the checking of snapshot validity.
+        """
+        for snap_data in self.dataList():
+            snap_data.setup(storage, ksdata, instClass)
+
+    def execute(self, storage, ksdata, instClass):
+        """ Create ThinLV snapshot.
+
+            The Blivet must be reseted before creation of the snapshot. This is
+            required because the storage could be changed in post section.
+        """
+        try_populate_devicetree(storage.devicetree)
+        for snap_data in self.dataList():
+            snap_data.execute(storage, ksdata, instClass)
+
+class SnapshotData(commands.snapshot.RHEL7_SnapshotData):
+    def __init__(self, *args, **kwargs):
+        commands.snapshot.RHEL7_SnapshotData.__init__(self, *args, **kwargs)
+        self.thin_snapshot = None
+
+    def setup(self, storage, ksdata, instClass):
+        """ Add ThinLV snapshot to Blivet model but do not create it.
+
+            This will plan snapshot creation on the end of the installation. This way
+            Blivet will do a validity checking for future snapshot.
+        """
+        log.debug("Snapshot name %s setup", self.name)
+        if not self.origin.count('/') == 1:
+            raise KickstartParseError(
+                        formatErrorMsg(self.lineno,
+                                       msg=_("Incorrectly specified origin of the snapshot."
+                                             " Use format \"VolGroup/LV_name\"")))
+        origin = self.origin.replace("/","-")
+        origin_dev = storage.devicetree.getDeviceByName(origin)
+
+        if not isinstance(origin_dev, LVMThinLogicalVolumeDevice):
+            raise KickstartParseError(
+                        formatErrorMsg(self.lineno,
+                                       msg=(_("Snapshot: origin \"%(origin)s\" of snapshot \"%(name)s\""
+                                              " is not a valid thin pool device.") %
+                                            {"origin": self.origin,
+                                             "name": self.name})))
+
+        self.thin_snapshot = None
+        try:
+            self.thin_snapshot = LVMThinSnapShotDevice(name=self.name,
+                                                       parents=[origin_dev.pool],
+                                                       segType="thin",
+                                                       origin=origin_dev)
+        except ValueError as e:
+            raise KickstartParseError(formatErrorMsg(self.lineno, msg=e))
+
+    def execute(self, storage, ksdata, instClass):
+        """ Register an action for snapshot creation.
+
+            Snapshot must be re-created because method blivet.reset() was called.
```

Comment:
I don't think ``blivet.reset()`` is being called. And if so, the ``self.thin_snapshot`` is hardly in the the devicetree.